src/lammpsio/snapshot.py

```diff
@@ -171,36 +230,128 @@ def to_hoomd_gsd(self, type_map=None):
-                for typeidx, typeid in enumerate(sorted_typeids):
-                    frame.particles.typeid[self.typeid == typeid] = typeidx
+                if self.type_label is not None:
+                    type_label_map = self.type_label
+                    frame.particles.types = type_label_map.types
+                    _set_type_id(
+                        self.typeid,
+                        frame.particles.typeid,
+                        type_label_map,
+                    )
+                else:
+                    sorted_typeids = numpy.sort(numpy.unique(self.typeid))
+                    frame.particles.types = [str(typeid) for typeid in sorted_typeids]
+                    for typeidx, typeid in enumerate(sorted_typeids):
+                        frame.particles.typeid[self.typeid == typeid] = typeidx
             else:
-                frame.particles.types = list(type_map.values())
-                reverse_type_map = {
-                    typeid: typeidx for typeidx, typeid in enumerate(type_map.keys())
-                }
-                for i, typeid in enumerate(self.typeid):
-                    frame.particles.typeid[i] = reverse_type_map[typeid]
+                warnings.warn(
+                    "type_map is deprecated, use label instead.",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
+                type_label_map = LabelMap(map=type_map)
+                frame.particles.types = type_label_map.types
+                _set_type_id(
+                    self.typeid,
+                    frame.particles.typeid,
+                    type_label_map,
+                )
         if self.has_charge():
             frame.particles.charge = self.charge.copy()
         if self.has_mass():
             frame.particles.mass = self.mass.copy()
         if self.has_molecule():
             frame.particles.body = self.molecule - 1
+        if self.has_bonds():
+            if self.bonds.has_typeid():
+                frame.bonds.N = self.bonds.N
+                frame.bonds.types = self.bonds.id
+                frame.bonds.group = self.bonds.members - 1
+                frame.bonds.typeid = numpy.zeros(self.bonds.N, dtype=int)
+                if self.bonds.type_label is not None:
+                    bond_label_map = self.bonds.type_label
+                else:
+                    sorted_typeids = numpy.sort(numpy.unique(self.bonds.typeid))
+                    bond_label_map = {}
+                    for typeid in sorted_typeids:
+                        bond_label_map[typeid] = str(typeid)
+                    bond_label_map = LabelMap(map=bond_label_map)
+                # set bond types & typeid
+                frame.bonds.types = bond_label_map.types
+                _set_type_id(
+                    self.bonds.typeid,
+                    frame.bonds.typeid,
+                    bond_label_map,
+                )
```

Comment:
I am a bit confused on this. 

Why are you setting the `frame.bonds.typeid` to zeros? Also this is setting `bond_label_map = self.bonds.type_label` and then overwriting it? 

And `_set_type_id` is a helper function that maps from LAMMPS type IDs to HOOMD type IDs but doesn't return anything. If it requires a bond label map as an input and doesn't modify it, I'm confused why it would also set it.

Reply:
> Why are you setting the `frame.bonds.typeid` to zeros?

I think your code requires that `frame.bonds.typeid` have something in it before you can call `_set_type_id` so this a temporary placeholder. It could also use `numpy.ones` or `numpy.empty`. You were filling `frame.bonds.typeid` with `self.bonds.id` above (i.e., 1, 2, ... N), which didn't make sense because that would imply each bond is its own type.
 
> Also this is setting `bond_label_map = self.bonds.type_label` and then overwriting it?
>
> And `_set_type_id` is a helper function that maps from LAMMPS type IDs to HOOMD type IDs but doesn't return anything. If it requires a bond label map as an input and doesn't modify it, I'm confused why it would also set it.

If `self.bonds.type_label is not None`, `bond_label_map` will be the label map, otherwise it will default to `None`. If `self.bonds.has_typeid() is True`, then `_set_type_id` will be called with `bond_label_map`. My comment above said that you should modify `_set_type_id` so that it creates the 1, 2, 3, ... default label map internally when it is given a label map that is None (to reduce code duplication) and returns the label map it actually used. That would mean that if `bond_label_map is not None`, the assignment does nothing; if it is None, it will return the auto-generated map. It is still possible for `bond_label_map` to be None if it was not set and there were no typeids.

Reply:
Ah I see, I had missed the bit about modifying `_set_type_id` to create `bond_label_map` if it is given `None`. That all makes sense! Thanks!